packages/phase/src/handler.ts

```diff
@@ -0,0 +1,118 @@
+// Copyright IBM Corp. 2014. All Rights Reserved.
+// Node module: loopback-phase
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {PhaseOptions} from './phase';
+const debug = require('debug')('loopback:phase:handler');
+
+// tslint:disable:no-any
+
+/**
+ * Context for the invocation of handlers by phase
+ */
+export interface Context {
+  [name: string]: any;
+  /**
+   * Error caught during the invocation
+   */
+  error?: any;
+}
+
+/**
+ * Handler function that takes the context
+ */
+export interface Handler {
+  /**
+   * @param ctx: Context
+   * @param chain: Optional handler chain if the handler is invoked as part of
+   * an invocation chain of handlers
+   */
+  (ctx: Context, chain?: HandlerChain): Promise<void> | void;
+}
+
+/**
+ * Handler chain that allows its handlers to control how to proceed:
+ *
+ * - next: call downstream handlers
+ * - return: mark the handler chain is done
+ * - throw: throw an error to abort the handler chain
+ */
+export class HandlerChain {
+  /**
+   * Indicate if the end of invocation chain is reached or the work is done
+   */
+  done?: boolean;
+  private index = 0;
+
+  constructor(private ctx: Context, private handlers: Handler[]) {
+    this.done = false;
+    Object.assign(this, ctx);
+  }
+
+  /**
+   * Run the downstream handlers
+   * @param ctx Optional context, default to `this.ctx`
+   */
+  async next(ctx: Context = this.ctx): Promise<void> {
+    debug('Dispatch to downstream handlers');
+    await asHandler(this.handlers.slice(++this.index))(ctx);
+    // Mark the context as `done` because the handler calls `next` to run
+    // downstream handlers
+    this.done = true;
+  }
+
+  /**
+   * Mark the context as `done` to stop the execution
+   */
+  return(): void {
+    debug('Return from the handler chain');
+    this.done = true;
+  }
+
+  /**
+   * Throw an error to abort the execution
+   * @param err
+   */
+  throw(err: any): void {
+    debug('Throw an error to abort the handler chain', err);
+    throw err;
+  }
+}
+
+/**
+ * Create a handler function that executes the given handlers sequentially
+ * @param handlers Array of handlers
+ */
+export function asHandler(
+  handlers: Handler[],
+  options: PhaseOptions = {},
+): Handler {
+  return async (ctx: Context) => {
+    if (options.parallel) {
+      debug('Executing handlers in parallel');
+      await Promise.all(handlers.map(h => h(ctx)));
```

Comment:
I don't think it's a good idea to run all handlers in parallel.

What if we have 3 handlers registered and the handler no.2 calls `ctx.return()`? It would be surprised to learn that the 3rd handler was executed to completion too!

What if we have 2 handlers registered and the first handler throws an error? I would find it surprising that the second handler was still executed. Also, are we going to wait until that second handler finishes, or will we return the error immediately?

Also, what's the point of `Phase#parallel` option when we are always running all handlers in parallel?

Please add tests to capture the expected behavior and verify the implementation.